papers/scipy/scipy.rst

```diff
@@ -0,0 +1,1706 @@
+    array([[  2.883e-05,   5.765e-05,   2.883e-05,
+              1.000e+00,  -1.349e+00,   4.602e-01],
+           [  1.000e+00,   2.000e+00,   1.000e+00,
+              1.000e+00,  -1.454e+00,   5.741e-01],
+           [  1.000e+00,   2.000e+00,   1.000e+00,
+              1.000e+00,  -1.681e+00,   8.198e-01]])
+
+
+**Limitations of the transfer function representation.**
+Earlier we said that the transfer function representation of
+moderate to large order IIR filters can result in numerical problems.
+Here we show an example.
+
+We consider the design of a Butterworth bandpass filter
+with order 10 with normalized pass band cutoff frequencies of 0.04
+and 0.16.::
+
+    >>> b, a = butter(10, [0.04, 0.16], btype="bandpass")
+
+We can compute the step response of this filter by applying it to
+an array of ones::
+
+    >>> x = np.ones(125)
+    >>> y = lfilter(b, a, x)
+    >>> plt.plot(y)
+
+The plot is shown in Figure :ref:`fig-unstable-butterworth`.
+(We haven't shown all the additional ``matplotlib`` function calls that
+we used to annotate the plot.)
+Clearly something is going wrong.
+
+.. figure:: figs/unstable_butterworth.pdf
+
+    Incorrect step response of the Butterworth bandpass filter of order
+    10 created using the transfer function representation.  Apparently the
+    filter is unstable--something has gone wrong with this representation.
+    :label:`fig-unstable-butterworth`
+
+We can try to determine the problem by checking the poles
+of the filter::
+
+    >>> z, p, k = tf2zpk(b, a)
+    >>> np.abs(p)
+    array([ 0.955,  0.955,  1.093,  1.093,  1.101,
+            1.052,  1.052,  0.879,  0.879,  0.969,
+            0.969,  0.836,  0.836,  0.788,  0.788,
+            0.744,  0.744,  0.725,  0.725,  0.723])
+
+The filter should have all poles inside the unit circle in the complex
+plane, but in this case five of the poles have magnitude greater than 1.
+This indicates a problem, which could be in the
+result returned by ``butter``, or in the conversion done
+by ``tf2zpk``.  The plot shown in Figure :ref:`fig-unstable-butterworth`
+makes clear that *something* is wrong with the coefficients in
+``b`` and ``a``.
+
+Let's design the same 10th order Butterworth filter as above,
+but in the SOS format::
+
+    >>> sos = butter(10, [0.04, 0.16],
+    ...              btype="bandpass", output="sos")
+
+In this case, all the poles are within the unit circle::
+
+    >>> z, p, k = sos2zpk(sos)
+    >>> np.abs(p)
+    array([ 0.788,  0.788,  0.8  ,  0.8  ,  0.818,
+            0.818,  0.854,  0.854,  0.877,  0.877,
+            0.903,  0.903,  0.936,  0.936,  0.955,
+            0.955,  0.964,  0.964,  0.988,  0.988])
+
+We can check the frequency response using ``scipy.signal.sosfreqz``::
+
+    >>> w, h = sosfreqz(sos, worN=8000)
+    >>> plt.plot(w/np.pi, np.abs(h))
+    [<matplotlib.lines.Line2D at 0x109ae9550>]
+    >>> plt.grid(alpha=0.25)
+    >>> plt.xlabel('Normalized frequency')
+    >>> plt.ylabel('Gain')
```

Comment:
Not sure if you did this on purpose, but this is half the plotting code for Fig 2 (there are 2 subplots in the paper). Perhaps confusing.

Reply:
As I mentioned in another comment, I think these transcripts from interactive sessions will be changed to show just the essential functon calls.